### ISSUE TITLE
Selection of topics for Vienna hackathon and Prague pre-hackathon

### ISSUE THREAD

**Issue by misaochan** [commons-app/apps-android-commons#353]
As the events are drawing nearer, I am attempting to consolidate the suggestions offered at #293 so that we can make a decision. The [Prague pre-hackathon](https://www.mediawiki.org/wiki/Wikimedia_Hackathon_2017/Prehackathon_Prague), based on my discussions with @VojtechDostal, would involve an introduction to the app to familiarize people with the codebase, and a small task related to the Nearby function. The [Vienna hackathon](https://www.mediawiki.org/wiki/Wikimedia_Hackathon_2017) would likely involve a larger task and would also provide an opportunity to collaborate with other WMF folks (e.g. newcomers, or the Wikipedia app devs) if needed.

**Potential tasks, in my opinion, based on suggestions received:**

Pre-hackathon:

- Displaying the Nearby results on a map
- Requesting Nearby images provided by the web-app that @tobias47n9e is working on
- Switching to a Wikidata query for Nearby results

Hackathon:

- Any of the tasks mentioned above
- Collaborating with other WMF developers to share knowledge and discuss how to reduce technical debt and other stalled issues (@whym has talked to Rachel Farrand and received encouraging feedback on this)
-  Recruiting and onboarding new contributors to the Commons app, especially those who are signing up as mentees for the hackathon

Thoughts?

**Comment by misaochan** [comment]
Once we have decided on our project for the hackathon, we will need to add it to https://phabricator.wikimedia.org/project/board/2530/ I believe.

**Comment by Abbe98** [comment]
As author of the [Articles by lat/lon without Images API](https://github.com/Abbe98/articles-by-lat-lon-without-images-service) I'm pinging in myself :-)

**Comment by misaochan** [comment]
Awesome, see you there @Abbe98 :)

**Comment by misaochan** [comment]
Received advice from Rachel Farrand via email:

> Besides creating the Phabricator task as you say (you will be surprised at how many people actually see these), a small number of people who want progress or help in advance of hackathons choose to email their project to the wikitech-l mailing list to get even more visibility and help.  That is totally up to you. :) 
> Also, at the opening of hackathon itself all participants will get the opportunity to announce what they are working on and what help (if any) they need for their projects. 
> 
> If you are looking to run an hour long session (for example something like "Brainstorming Session on Improving the Wikimedia Commons App" or "Wikimedia Commons App: What we have now and where to go next" or whatever) at the hackathon in one of the breakout rooms instructions on how to schedule those will be posted much closer to the event as well. You don't have to worry about being able to get a space during the event, we wont run out of space for you. 
> 
> If there are any WMF developers that you want to coordinate with in advance of the event I am also happy to introduce you by email or if you like I can check and see which WMF mobile devs will be attending. 
> 
> Just let me and Srishti know what you need in order for your projects to be successful and we will do our best to help! 

I think the hour-long brainstorming session that she mentioned would be a good addition to our plans, especially as it is an opportunity to get feedback from a wide range of fellow developers.

Are there any WMF devs in particular whom we would like to invite/request for collaboration during the event?

**Comment by tobias47n9e** [comment]
I created a task for the discussion: https://phabricator.wikimedia.org/T158994
A task for the image request / campaign request project: https://phabricator.wikimedia.org/T158995
And a task for the Django/MW-OAuth intro: https://phabricator.wikimedia.org/T158998

@misaochan Can you create the project taks for the Android app?

**Comment by misaochan** [comment]
Hi @tobias47n9e ,

Just clarifying how this works: We want to create a Phab task for each task that we plan on working on for the hackathon, right? But what if we haven't decided exactly which task to work on? I think doing all of the tasks that I listed above is a bit too much, even if they are split between the pre-hackathon and hackathon. But I'm hoping to get more feedback from folks who are interested before we settle on the final task(s).

Or can I just start one general "task" for our app and list the possible projects within that task?

Tagging @VojtechDostal to ask about how WMCZ would like the prehackathon to be organized.

**Comment by VojtechDostal** [comment]
Hi all! Thanks for working on this. From my perspective it would be best to list all possible tasks we might work on and prioritize those. We may decide to work on some of these in the weeks before the events, while additional decisions can be made during the events themselves. 

WMCZ will mostly be behind logistics of both events and won't decide which tasks are "appropriate" and which are not. Our incentive is to improve the "Nearby" function and add Wikidata as a source for those data but generally we'll support all that sounds reasonable and is within the scope of the Wikimedia Commons app :-)

**Comment by tobias47n9e** [comment]
I would say we create task in "Projects" (https://phabricator.wikimedia.org/project/board/2530/) to show people that they can work on this project during the hackathon. They can also join our discussion session. We could also add one item in skill share if we want to show novice developers how they can tackle and issue with the app. I think we should stick with Github for subtasks, as it would only duplicate a lot of discussions.

**Comment by misaochan** [comment]
@tobias47n9e done. :)

What do you think the length of a skill share session should be, and how should it be structured? Maybe fixing a minor bug, so a couple hours all around including introduction and helping them set up the development environment?

**Comment by tobias47n9e** [comment]
@misaochan I think that is up to you and the other Android developers. Maybe two people could do one hour each, so they don't exhaust themselves with explaining stuff :)

**Comment by whym** [comment]
How about asking participants to try setting up the developing environment at home? If they can successfully install Android Studio etc at home, that's great. Even if it's only partially done, it could save some waiting time (e.g. for downloading things) at the hackathon.

Also, is there a way to know the level of familiarity to Android/Java of the participants, perhaps by asking at the online registration or in a follow up email? That would help planning things.

**Comment by nicolas-raoul** [comment]
Just to say that I agree: When doing trainings it is easy to lose a lot of
time with silly things like Wi-Fi connection or environment problems, so
indeed I would recommend to ask participants to kindly:
- install Android Studio, or at least download it
- install Git
- bring a USB cable
- bring various Android devices
Having the installers on a USB stick can help for people who for any reason
have trouble downloading it.

**Comment by VojtechDostal** [comment]
@whym We will ask about level of familiarity to Android/Java/GitHub in a registration form that we plan to send out with @misaochan in a few days. Thanks for comments all!

**Comment by misaochan** [comment]
Good idea, I'll add those requests to the pre-hackathon wiki page. Shall I also add them to the Phab task for our project in the hackathon, or will we also have a page for that?

**Comment by misaochan** [comment]
Added to the [pre-hackathon wiki](https://www.mediawiki.org/wiki/Wikimedia_Hackathon_2017/Prehackathon_Prague), do let me know if I missed anything.

**Comment by misaochan** [comment]
Registration and scholarship applications are open! Apply [here](https://docs.google.com/forms/d/e/1FAIpQLScvPa8LIz_YTOIcToaAAkAvnZ5OoJ2F5YUlfelDcTcutL24cg/viewform). 

Tagging those who have made contributions in the past: @whym @domdomegg @veyndan @tobias47n9e  @anirudh24seven @AtaZh @jayvdb . Please feel free to tag anyone else whom you think might be interested, thanks!

**Comment by domdomegg** [comment]
Greatly appreciate the offer but unfortunately I'm busy then

**Comment by asadsalman** [comment]
Do you guys think it would be better to have topics that people can work on during both the pre-hackathon _and_ the main hackathon event? Or should the work done at both events be kept separate?

Moreover, are you guys looking to have "Skill Shares" like the main event has?

**Comment by misaochan** [comment]
@asadsalman I think it might be a good idea to have one task for the pre-hackathon and another task for the hackathon initially, and see how it goes. If the pre-hackathon task finishes early then we can start on the hackathon task and add more to the list, or if it doesn't finish in time then we can continue during the hackathon? What do the others think?

Could you elaborate on the "Skill Share" sessions please? From what I'm seeing, they look like tutorials/introductory sessions? In that case, if there are newbies present at the pre-hackathon, I will probably run one during the first day to get everyone up to speed - just a basic introduction on the app, how it works, how to build, etc. If everyone at the pre-hackathon is experienced, anyone with a specific niche is welcome to conduct a session. :) I think @tobias47n9e has offered to do a short intro on the web app in any case.

**Comment by maskaravivek** [comment]
The idea of having some sort of a "Skill share" session definitely seems good. If we are doing some form of a short intro web app we can decide up on the tasks and I would love to do some work in setting it up. @tobias47n9e 

I would also love to volunteer in helping the newbies setup things as mentioned by @nicolas-raoul at the start of the pre-hackathon.

**Comment by tobias47n9e** [comment]
I just wanted to say that we should ensure that all of the things we do during the hackathon are also documented here for those that can't attend. Although there is budget for some people to attend it will most likely impossible to invite everyone (for those that still are active next year we can probably figure out a scholarship rotation).

I already finished my skill-share tutorial here: https://wikitech.wikimedia.org/wiki/Help:Tool_Labs/My_first_Django_OAuth_tool So if anyone wants to go through the steps and see what can be improved they are very welcome.

**Comment by domdomegg** [comment]
Do you think it might be possible to record some of it, especially the introductory sessions? Doesn't have to be v. high quality just a phone strapped to a makeshift tripod + a half-decent mic would probs be good enough.

**Comment by Akanksha08** [comment]
Hello folks! I am really interested in participating in the pre hackathon as well as the hackathon. I am done with the pre preparation mentioned here: [](https://www.mediawiki.org/wiki/Wikimedia_Hackathon_2017/Prehackathon_Prague)

Please guide me how do I proceed. I would like to get some hands on before I attend the actual hackathon.

**Comment by misaochan** [comment]
I think we will probably have to ask @VojtechDostal about the logistics, as I'm not sure what recording etc facilities are available. Also, will we want slides for the introductory session? I figure they would be easier for people to refer to, and I'd be happy to make them, but will there be a projector in the room that we will use?

@Akanksha08 welcome! :) Please take a look through our Issues section, especially those tagged [beginner-friendly](https://github.com/commons-app/apps-android-commons/issues?q=is%3Aopen+is%3Aissue+label%3A%22beginner+friendly%22). Now that you have set up the dev environment and built the project, the best way to learn IMO is to try to tackle a small issue. If you encounter any problems or can't find an issue that you are interested in working on, please let us know.

**Comment by misaochan** [comment]
Potential plans in a bit more detail:

**Pre-hackathon:**

- Introductory/skill share sessions
- Tasks to work on: Those who prefer Android dev can work on the Android app, and those who prefer web dev can work on the web app. @tobias47n9e , is my interpretation of your plans for the web app correct?
 - (1) Switching to a Wikidata query for Nearby results in the Android app
 - (2) [Wmflabs page to request images and plan campaigns](https://phabricator.wikimedia.org/T158995)

**Hackathon:**

- If task (1) is not yet complete, work on that. Otherwise, work on (3) Displaying the Nearby results on a map (Android app). Web dev tasks will be Tobias's choice.
- Open discussion: [Commons app - Current state and future plans](https://phabricator.wikimedia.org/T158994) 
 - Possible discussion with Wikipedia Android app devs to share knowledge and discuss how to reduce technical debt and other stalled issues. (Hi @niedzielski , sorry for tagging you out of the blue! :) Do you think this might be possible?)
- Recruiting and onboarding new contributors to the Commons app, especially those who are signing up as mentees for the hackathon

**Comment by niedzielski** [comment]
Sorry for the ultra slow response :] I won't be at the hackathon unfortunately but @dbrant and @mdholloway might. I can probably also do a Hangout if that would be useful.

**Comment by dbrant** [comment]
@misaochan I will be at the hackathon, and will likely have some bandwidth to help with this, either in the form of actual hacking or product/technical guidance.

**Comment by misaochan** [comment]
Thanks guys! :) @dbrant that would be great, thank you. Shall we arrange a time with you closer to the date of the hackathon, or would you like to just drop by?

**Comment by dbrant** [comment]
Sure, it would probably make sense to meet briefly before the hackathon to solidify how I can be of service. (or just discuss it on the phab task)

**Comment by misaochan** [comment]
I've taken the liberty of setting up a rough programme structure for the prehackathon based on discussion with @VojtechDostal . If I'm missing anything or anyone disagrees with anything, please let me know.

**Comment by misaochan** [comment]
@whym - If I understand correctly, you will be attending, yeah? Would you like to talk briefly about the Commons app [stats tool](http://tools.wmflabs.org/commons-app-stats/) that you have created?

**Comment by whym** [comment]
I'm not a great speaker but sure. It's just a simple tool, so I think the talk will be (and should be) a quick one.

**Comment by misaochan** [comment]
Sounds good to me, have added you in. :)

**Comment by misaochan** [comment]
So it seems like we are close to having the minimum viable product ready for the switch to Wikidata. :) In that case, I figure we should tentatively switch the main task for the pre-hackathon and hackathon (Android track) to (3) Displaying the Nearby results on a map (Android app) #334.

Additionally, we could work on any of the other issues that we are tracking, especially relevant ones like #502 , #508 etc.

**Comment by nicolas-raoul** [comment]
May I suggest #252 ? (Allow people to upload a picture for a given nearby item) so that people don't have to go edit Wikidata to make the item disappear.

**Comment by misaochan** [comment]
Yeah, that would be nice, perhaps also #278

**Comment by misaochan** [comment]
Being able to fix some of the high priority issues (albeit not necessarily Nearby-related) would be good too. #510 currently limits our userbase to those with mid-range phones and above I think, which is a limitation that we should try to remove.

**Comment by whym** [comment]
> Would you like to talk briefly about the Commons app stats tool that you have created?

I've noticed that I might miss the opening by one or two hours. At best (and if the passport control is really fast for me) I might arrive just in time, but I'm not sure about that. @misaochan Would you like to talk about the stats tool for me, if I am not yet present by then? I can send slides beforehand.

**Comment by misaochan** [comment]
@whym I think it would be okay for your talk to be at a different time (perhaps after lunch). It might be a good idea to break up the talks anyway, and it isn't essential that people know about the stats tool before starting work. :)

**Comment by misaochan** [comment]
@maskaravivek I noted that on the registration form you offered to help the newcomers with getting set up - that would be a great help, thanks! Can you be on hand for that after the introductory session? I will be helping as well, but it would be good for two of us in case there is a substantial number of newcomers (not sure if there will be, but one can always hope :D)

**Comment by whym** [comment]
Ah right, after lunch sounds good to me.

**Comment by misaochan** [comment]
@whym If you manage to arrive on time, would you like to help me talk about the Nearby code? Since the latest incarnation of it is largely yours. No need to make any slides or anything for that one, we can probably just pull up the IDE on the projected laptop. I'll try and do what I can if you're not around, but if you are then you could probably clarify some things that I might miss.

**Comment by whym** [comment]
> If you manage to arrive on time, would you like to help me talk about the Nearby code?

Sure, I'm happy to help. My knowledge tends to be limited to specific classes and methods I touched, though. I hope I can learn more about the classes I haven't touched and the overall structure during the hackathon.

**Comment by misaochan** [comment]
Final draft of program and selected tasks is up. As usual, feel free to discuss if you disagree with anything.

Also, please feel free to work on any of the tasks mentioned there before the pre-hackathon, especially those pertaining to bug fixes. We do not need to reserve all of them for the pre-hackathon/hackathon, if anything has already been fixed we can simply move on down the list. :)

**Comment by misaochan** [comment]
For the onboarding session during the hackathon, I was thinking that, after getting them set up, we could pick a few beginner-friendly tasks and let them try it on their own, helping them where needed? Would that be a good idea?

**Comment by maskaravivek** [comment]
@misaochan Sure, would love to help newcomers with that. :)
